### DISCUSSION TITLE
Any Alternatives to GStreamer?

### DISCUSSION THREAD

**Discussion by MIDIculous** [ant-media/Ant-Media-Server#6635] [General]
My developer and I have been spending more time fixing GStreamer than we do actually implementing the features. We are using Gstreamer in order to have control over the video and audio stream so we can record locally. 

Is anyone using anything other than GStreamer or FFMPEG for more enhanced video control with overlays, and text etc...?

**Comment by asharamseervi** [comment]
Hi @MIDIculous,

Thanks for asking it.

OBS Studio can be a strong alternative to GStreamer, especially if you want more intuitive control over video and audio streams, recording to local is also available out-of-the-box.

OBS is also community-driven and provides intuitive controls for streaming along with multiple added features such as overlays, etc.

And, It's widely adopted streaming industry and I believe, it'll serve your purpose well. Surely, worth exploring!

Let us know if you've any more questions.

**Reply by MIDIculous** [reply]
@asharamseervi 
Perhaps you are not understanding. We are looking for an alternative media player for this type of scenario: https://antmedia.io/gstreamer-tutorial-how-to-publish-play-webrtc-streams/

We are looking for something different for streaming WebRtc but with more control.

**Reply by klasadoroty** [reply]
hm.. It turns out, however, that you did not understand - if you want more intuitive control over video and audio streams, then OBS is what you are looking for. Both solutions are free. 
If neither of the above, then I would suggest investing in paid software.

**Reply by MIDIculous** [reply]
> hm.. It turns out, however, that you did not understand - if you want more intuitive control over video and audio streams, then OBS is what you are looking for. Both solutions are free. If neither of the above, then I would suggest investing in paid software.

I am building a WebRTC desktop application in C++. I cannot build a desktop application with OBS!

**Reply by klasadoroty** [reply]
Ok. this changes everything.
You need to be more precise in your writing. Read your first post.

Then OBS is out.

**Reply by asharamseervi** [reply]
Apologies for misunderstanding your question, @MIDIculous. Thank you, @klasadoroty, for taking over.

**Comment by MIDIculous** [comment]
Also we are creating a C++ App to do the WebRTC stuff. We just need additional audio and video control and GStreamer was suggested. Sonia there anything close to GStreamer for video capture and WebRTC that is more intuitive.

**Comment by rahul78275** [comment]
Hello @MIDIculous 

Thank you for asking the question. If you elaborate more about your use-case what you want to achieve, you can get better advice from members.

May I know your interest for choosing C++ for WebRTC stuff ? You can make use of **[electronjs](https://www.electronjs.org/)** for creating desktop application.

**Comment by MIDIculous** [comment]
### Background
We have an existing application that we have already produced and it's in C++ using the Juce cross-platform framework. We want to now implement  Audio, Video and MIDI into our application.
- We are able to handle the MIDI over WebRTC Data Channel with no problem using (https://github.com/paullouisageneau/libdatachannel)
- We are also able to handle basic Video and Audio using any of the off the shelf tools for WebRTC. 

### The Issue
The issue lies in the fact that we don't just want to do basic Video/Audio stuff. We would like to do things like:
- Image overlays
- Text Overlays
- Audio and Video Recording to the desktop

### AntMedia's Solution
So the solution suggested as per [AntMedia's Website](https://antmedia.io/gstreamer-tutorial-how-to-publish-play-webrtc-streams/) is to use GStreamer, but we are finding that we are spending more time fixing Gstreamer's bugs and workarounds then actually implementing it.

### Question
- Are there any other solutions for rich media streaming with overlays, text, and desktop recording other than GStreamer?
- We would prefer C++ and cross-platform as our app is already in C++

**Reply by rahul7827** [reply]
Hello @MIDIculous 

I am trying to help you as much I can, might be my answer is not useful for you because I have no expertise in C++.

In antmedia solution, There is a [media push plugin](https://github.com/ant-media/Plugins/tree/master/MediaPushPlugin) using that you can achieve overlays. 

Recording feature is also in antmedia server, But it doesn't record on your local system. You can make use of [antmedia api](https://antmedia.io/rest/). Once recording file is available you can download the file by your script.

**Hack:** You can make use of datachannel for overlays over player screen. If player is under your control. [Reference](https://github.com/ant-media/Ant-Media-Server/issues/6545#issuecomment-2278392998)

**Reply by MIDIculous** [reply]
@rahul78275 Thanks so much for the response. It is as I have guessed it: I need to do custom development in order to do what I really neeed to do.

**Reply by amarantmeida** [reply]
@MIDIculous As suggested by @rahul7827, [Media push plugin](https://github.com/ant-media/Plugins/tree/master/MediaPushPlugin) can be used to achieve overlays. Please let us know if you need any other information from our side.
